src/typecheck.rs

```diff
@@ -1499,17 +1492,18 @@ fn instantiate_foralls(state: &mut State, mut ty: TypeWrapper, inst: ForallInst)
 }
 
 /// Type of unary operations.
-pub fn get_uop_type(state: &mut State, op: &UnaryOp) -> Result<TypeWrapper, TypecheckError> {
+pub fn get_uop_type(
+    state: &mut State,
+    op: &UnaryOp,
+) -> Result<(TypeWrapper, TypeWrapper), TypecheckError> {
     Ok(match op {
         // forall a. bool -> a -> a -> a
         UnaryOp::Ite() => {
             let branches = TypeWrapper::Ptr(new_var(state.table));
 
-            mk_tyw_arrow!(
-                AbsType::Bool(),
-                branches.clone(),
-                branches.clone(),
-                branches
+            (
+                mk_typewrapper::bool(),
+                mk_tyw_arrow!(branches.clone(), branches.clone(), branches),
```

Comment:
I'm a bit curious: I was expecting that given that you still return an arrow type here, you could still exhibit the issue given a well-crafted `if-then-else` statement (something like `let x : Num = if true then 1 else "foo"`), but I couldn't get the typechecker to issue a bad error message. Is there a reason for that?

Reply:
You're right that in general, this kind of operator - the ones that are "lying" about their arity because they accept additional arguments lazily, and thus needs to return an arrow type - should still exhibit the issue. Here, you tried `(ifthenelse true) 1 "foo"` where `(ifthenelse true) : a -> a -> a`.

As long as the type mismatch is in the argument, the error should be local, because of the way of typechecking function application work: we first first determine the type of the function, then try to match the argument against the expected type.

However, having a wrong outer annotation while the if-then-else itself is well-typed should trigger the problem:

```
nickel> (if true then 0 else 0) : Str
error: Incompatible types
  ┌─ repl-input-4:1:15
  │
1 │ (if true then 0 else 0) : Str
  │               ^ this expression
  │
  = The type of the expression was expected to be `Str`
  = The type of the expression was inferred to be `Num`
  = These types are not compatible
```

Hmm, it doesn't...After more thought, this is due to the order in which the unification is done, which was a bit different in the removed code versus the code for standard application (which it shouldn't have been, really). Because of row types (see comment https://github.com/tweag/nickel/blob/9a594f460008c467d4371259ce2999621b847720/src/typecheck.rs#L564), the left hand side of an application is checked first before the argument. When typechecking `(ifthenelse true) 0 0 : Str`:
1. On the first application, it first typechecks the left hand side: generates a type `a -> Str`, typecheck `(ifthenelse true) 0` against it
2. Again, generate a type `b -> a -> Str`, typecheck `(ifthenelse true)` against it
3. By `get_uop_type`, `ifthenelse true` has type `c -> c -> c`. Unifying the two sets the type to `Str -> Str -> Str`.
4. Then, typing the argument of the inner application raise an error: `Num != Str`.

On the other hand, if the type is not polymorphic, you do get an arrow type error, because at step `3` you try to unify the whole `Num -> Num -> Num` with `b -> a -> Str`, which fails:

```
nickel> let ite : Bool -> Num -> Num -> Num = fun x y z => z  in ((ite true 0 0) : Str) 
error: Function types mismatch
  ┌─ :1:23
  │
1 │ Bool -> Num -> Num -> Str
  │                       --- This part of the expected type
  │
  ┌─ :1:23
  │
1 │ Bool -> Num -> Num -> Num
  │                       --- does not match this part of the inferred type
  │
  ┌─ repl-input-3:1:60
  │
1 │ let ite : Bool -> Num -> Num -> Num = fun x y z => z  in ((ite true 0 0) : Str) 
  │                                                            ^^^ this expression
  │
  = The type of the expression was expected to be `Bool -> Num -> Num -> Str`
  = The type of the expression was inferred to be `Bool -> Num -> Num -> Num`
  = Could not match the two function types
```

These "Arrow type mismatch" are somehow an error reporting feature. Beforehand, both errors would be reported the same, but this was confusing in the case where you explicitly define `let ite : Bool -> Num -> Num -> Num`, because this got you an `expected Str, got Num` located on the `ite` (IIRC), so more context about the "top-level" unification that was actually going on was added. But at the end of the day, the error is caused by unification on some sub-component.

Finally, error reporting aside, it doesn't really make sense to return an arrow type for primitive operators, just to generate fresh variables to unify with the components of these arrows: in some way, we were constructing an arrow that we generically destructed immediately, all on the type system side, which sounds wasteful.